### PR TITLE
add an article about finding the cause of a memory leak in jest tests

### PR DESCRIPTION
**Pull request by lukasmlady**
See Paper doc for an interactive/commentable version of the post at https://paper.dropbox.com/doc/Blog-post-Finding-the-cause-of-a-memory-leak-in-Jest-tests--A2aPrFM5xkwhoPrkansPey0LAg-fiKgnvTeJznfRCrAgGERD.

/cc @closeio/engineering 

Snímek obrazovky 2020-06-24 v 11 31 16
Snímek obrazovky 2020-06-24 v 11 31 31
Snímek obrazovky 2020-06-24 v 11 31 42

### CODE REVIEW

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+title: 'Finding the cause of a memory leak in Jest tests'
+date: 2020-06-24
+permalink: posts/finding-the-cause-of-a-memory-leak-in-jest
+author: Lukáš Mladý
+---
+
+As we’ve been increasing our test coverage using [Jest](https://jestjs.io/) and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro), we have started seeing our CI-run tests occasionally failing with “out of memory” errors.
+
+```
+    FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```
Comment (philfreo) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:12]: here and below, it'll look better if you remove the indents

Comment (lukasmlady) [reply _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md]: thanks for spotting that, I wondered why was it indented but disregarded it as "styling"… fixed in 7ffe3fc1dad9ca73b0f5a3efb3970f9e86328c05

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+But these happened _only in the CI_ and _only occasionally_.
+
+We've managed to fix the problem after a couple of hours, and we’d like to share the process of how we found the cause of a memory leak. Read on!
+
+## Running tests with heap usage recording
+
+We've tried increasing memory allocation to see if the tests just take too much memory, but in a steady manner. That did not work which showed us that we’ve got a memory leak in our tests.
+
+When looking around for advice, we’ve found an article about [Jest memory leaks](https://chanind.github.io/javascript/2019/10/12/jest-tests-memory-leak.html) and went ahead and ran the following command:
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:23]: I'd change both these `we've` to just `we`

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+We went ahead and brought in Chrome to inspect the node process that runs the tests:
+
+```
+node --inspect-brk --expose-gc ./node_modules/.bin/jest --runInBand --logHeapUsage
+```
+
+Then go to Chrome -> enter `chrome://inspect` and connect the debugger.
+
+We took 3 heap snapshots, compared what was increasing disproportionally (array constructor) and dug deeper.
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:65]: `disproportionately`

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+[![Heap Snapshot 3](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3.png)](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3.png)
+
+There was a series of same-memory-allocation blocks (see the **Shallow Size** column in the screenshot below) that was growing for each snapshot. We drilled into those blocks and found traces of `localStorage` and its mocking. The `localStorageSetItemSpy` in the screenshot below looked like something _we_ wrote.
+
+[![Heap Snapshot 3 — Drilled-in](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3-drilled-in.png)](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3-drilled-in.png)
+
+## The root cause & solution
+
+When `grep`-ing for `localStorageSetItemSpy`, we got a couple of test files having that reference.
+In one of those files, there was a missing `.mockRestore` call (note that `.mockReset` only resets call counts etc.):
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:78]: perhaps link `mockRestore` and `mockReset` to their respective Jest docs?

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+[![Heap Snapshot 2](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-2.png)](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-2.png)
+[![Heap Snapshot 3](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3.png)](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3.png)
+
+There was a series of same-memory-allocation blocks (see the **Shallow Size** column in the screenshot below) that was growing for each snapshot. We drilled into those blocks and found traces of `localStorage` and its mocking. The `localStorageSetItemSpy` in the screenshot below looked like something _we_ wrote.
+
+[![Heap Snapshot 3 — Drilled-in](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3-drilled-in.png)](/assets/finding-the-cause-of-a-memory-leak-in-jest/heap-snapshot-3-drilled-in.png)
+
+## The root cause & solution
+
+When `grep`-ing for `localStorageSetItemSpy`, we got a couple of test files having that reference.
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:77]: I'd suggest just ``We found a couple of test files referencing `localStorageSetItemSpy``, that removes some passive languange

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+});
+```
+
+We suspect this mock was carried over to other tests as it was touching on a global object's prototype.
+
+## Closing thoughts
+
+Memory leaks can be nasty and pretty hard to find. We found that the hard way. We should always make sure to clean up after ourselves in our tests when mocking, even though Jest runs each test file in a separate process.
+
+We hope the process for finding the cause of a memory leak outlined in this article will help you fix issues like that in your projects.
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:113]: `will help you fix similar issues in your projects.`

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
+---
+layout: post
+title: 'Finding the cause of a memory leak in Jest tests'
+date: 2020-06-24
+permalink: posts/finding-the-cause-of-a-memory-leak-in-jest
+author: Lukáš Mladý
+---
+
+As we’ve been increasing our test coverage using [Jest](https://jestjs.io/) and [React Testing Library](https://testing-library.com/docs/react-testing-library/intro), we have started seeing our CI-run tests occasionally failing with “out of memory” errors.
```
Comment (essmahr) [inline _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md:9]: inconsistency with `we've`/`we have` (I'd maybe change the second to just `we started seeing...`)

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md (hunk 6 above)
Comment (lukasmlady) [reply _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md]: Fixed in b9798ef

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md (hunk 4 above)
Comment (lukasmlady) [reply _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md]: Fixed in b9798ef

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md (hunk 3 above)
Comment (lukasmlady) [reply _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md]: Fixed in b9798ef

File: _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md (hunk 2 above)
Comment (lukasmlady) [reply _posts/2020-06-24-finding-the-cause-of-a-memory-leak-in-jest.md]: Fixed in b9798ef

**Comment by lukasmlady** [comment]
Thanks @philfreo, @essmahr and @baseten for your feedback!

**Pull request merged**